### PR TITLE
User module for Sybil resistance

### LINKED ISSUE #162: Spec out user module

### LINKED ISSUE #182: User vouch/invite module for Sybil resistance
Implementation of #162 on top of #181.

### PR DESCRIPTION
**Pull request by mohitmamoria**
Fixes #162. Fixes #182.

### CODE REVIEW

File: x/user/spec/README.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+### Parameters
+
+Each of these parameters can be voted on by governance, but will be assigned sane defaults for genesis.
+
+```go
+type Params struct {
+    // minimum rewards to be earned to unlock the right to vouch
+    // it is an array because on networks like Reddit's,
+    // the threshold can be defined as a combination of multiple coins
+    ThresholdAmount     []sdk.Coin
```
Comment (shanev) [inline x/user/spec/README.md:32]: This can be `sdk.Coins`

**Review by shanev — COMMENTED** [review]
This is great! Very similar to the design I was thinking about. Like the flexibility of the Comment field. It's a bit like a memo field that can be used for custom verification methods in the future.

File: x/user/spec/README.md (hunk 1 above)
Comment (mohitmamoria) [reply x/user/spec/README.md]: Changed.

**Comment by shanev** [comment]
The last thing to finish this up is to add what events to fire. Stakewatcher can use this to index each person's vouch/invite list.

File: x/user/client/cli/tx.go [hunk 2]
```diff
@@ -29,23 +29,22 @@ func GetTxCmd(cdc *codec.Codec) *cobra.Command {
...
-func GetCmdPost(cdc *codec.Codec) *cobra.Command {
+// GetCmdVouch implements the delegate command.
+func GetCmdVouch(cdc *codec.Codec) *cobra.Command {
 	return &cobra.Command{
-		Use:   "post [vendor-id] [post-id] [body] [reward_address]",
-		Args:  cobra.MinimumNArgs(3),
-		Short: "Register a post",
+		Use:   "vouch [vouched] [comment]",
+		Args:  cobra.MinimumNArgs(1),
+		Short: "Vouch a user",
```
Comment (shanev) [inline x/user/client/cli/tx.go:43]: Maybe "Vouch for a user"?

File: x/user/keeper/vouch.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	vouchedKey := types.VouchedKey(vouched)
+	voucherKey := types.VoucherKey(voucher, vouched)
+	value := k.cdc.MustMarshalBinaryBare(&vouch)
+	store.Set(vouchedKey, value)
+	store.Set(voucherKey, value)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypePost,
+			// sdk.NewAttribute(types.AttributeKeyVendorID, fmt.Sprintf("%d", vendorID)),
```
Comment (shanev) [inline x/user/keeper/vouch.go:98]: We can add attributes for the vouched, voucher, and sender (same as voucher).

File: x/user/spec/README.md [hunk 4]
```diff
@@ -1,20 +1,22 @@
...
+Based on the above facts, the inspiration has been derived to design Sybil-resistance mechanism like the following:
 
-Moderate transactions are for signaling that a post contains spam or illegal content. Moderated posts immediately go to governance for adjudication. Curation is paused during this time. For posts deemed as inappropriate, curators get slashed and the moderator gets a reward. For posts not deemed as inappropriate, curation resumes while the moderator gets slashed.
+- Except for the first set of users (Genesis users), a new user can only start earning reward once vouched for by a previous user.
+- To earn the right to vouch (RtV) a user, the user must do some work (i.e. earn a minimum of x reward).
 
-Only data for posts currently being curated are stored on-chain. Everything else is deleted after the curation window completes and rewards are paid out.
+Note: If multiple users vouch for the same new user, only the first voucher's vouch would be considered, and the others' vouches would be left unused.
```
Comment (shanev) [inline x/user/spec/README.md:19]: Good for now. Possible to incentivize in the future :)

**Review by shanev — APPROVED** [review]
Looks great! Really well architected.  🌈🏅

At some point later we would add a querier to check if a user has been vouched for or not.

**Pull request merged**